tensorflow/contrib/ignite/kernels/ignite_dataset_ops.cc

```diff
@@ -0,0 +1,127 @@
+            std::string cache = "";
+            OP_REQUIRES_OK(ctx, ParseScalarArgument<std::string>(ctx, "cache", &cache));
+            *output = new Dataset(ctx, cache);
+        }
+
+    private:
+        class Dataset : public GraphDatasetBase {
+        private:
+            Ignite grid;
+            const std::string cacheName;
+        public:
+            Dataset(OpKernelContext *ctx, const string &cache)
+                    : GraphDatasetBase(ctx), cacheName(cache) {
+                grid = IgniteHolder::Instance().getIgnite();
+            }
+
+            std::unique_ptr<IteratorBase> MakeIteratorInternal(
+                    const string &prefix) const override {
+                return std::unique_ptr<IteratorBase>(
+                        new Iterator({this, strings::StrCat(prefix, "::Ignite")}));
+            }
+
+            const DataTypeVector &output_dtypes() const override {
+                static DataTypeVector *dtypes = new DataTypeVector({DT_STRING});
+                return *dtypes;
+            }
+
+            const std::vector<PartialTensorShape> &output_shapes() const override {
+                static std::vector<PartialTensorShape> *shapes =
+                        new std::vector<PartialTensorShape>({{}});
+                return *shapes;
+            }
+
+            string DebugString() const override { return "IgniteDatasetOp::Dataset"; }
+
+        protected:
+            Status AsGraphDefInternal(DatasetGraphDefBuilder *b,
+                                      Node **output) const override {
+                return Status::OK();
+            }
+
+        private:
+            class Iterator : public DatasetIterator<Dataset> {
+                typedef std::vector<CacheEntry<int, std::string> > ResVector;
+            private:
+                ResVector::const_iterator iter;
+                ResVector::const_iterator iter_end;
+                ResVector res;
+            public:
+                explicit Iterator(const Params &params) : DatasetIterator<Dataset>(params) {
+                    ScanQuery scan;
+
+                    const char *cacheName = params.dataset->cacheName.c_str();
+
+                    Ignite grid = params.dataset->grid;
+                    Cache<int, std::string> cac = grid.GetCache<int, std::string>(cacheName);
+
+                    cac.Query(scan).GetAll(res);
+
+                    iter = res.begin();
+                    iter_end = res.end();
+                }
+
+                Status GetNextInternal(IteratorContext *ctx,
+                                       std::vector<Tensor> *out_tensors,
+                                       bool *end_of_sequence) override {
+                    Tensor line_tensor(cpu_allocator(), DT_STRING, {});
+
+                    if (iter != iter_end) {
+                        std::string val = iter->GetValue();
+                        line_tensor.scalar<string>()() = val;
```

Comment:
Use `std::move(val)` here to avoid a copy.

Reply:
Fixed.